src/Language/DifferentialDatalog/CompileJava.hs

```diff
@@ -0,0 +1,487 @@
+
+-- convert a simple type to a Java type
+simpleType :: Type -> Doc
+simpleType TBool{..} = "boolean"
+simpleType TInt{..} = "BigInt"
+simpleType TString{..} = "String"
+simpleType TSigned{..} | typeWidth == 8  = "byte"
+simpleType TSigned{..} | typeWidth == 16 = "short"
+simpleType TSigned{..} | typeWidth == 32 = "int"
+simpleType TSigned{..} | typeWidth == 64 = "long"
+-- the mapping of unsigned values is not really correct if the value overflows
+-- but Java does not really support unsigned types, so any mapping will be a compromise
+simpleType TBit{..}    | typeWidth == 8  = "byte"
+simpleType TBit{..}    | typeWidth == 16 = "short"
+simpleType TBit{..}    | typeWidth == 32 = "int"
+simpleType TBit{..}    | typeWidth == 64 = "long"
+simpleType TStruct{..} | length typeCons == 1 = pp $ capitalize $ consName $ head typeCons
+simpleType t = error $ "Unsupported type " ++ show t
+
+-- given a simple type returns the accessor for a DDlogRecord to extract a field of this type
+getRecord :: Type -> Doc
+getRecord TBool{..}   = "getBoolean()"
+getRecord TInt{..}    = "getU128()"
+getRecord TString{..} = "getString()"
+getRecord TSigned{..} = "getLong()"
+getRecord TBit{..}    = "getLong()"
+getRecord t = error $ "Unsupported type " ++ show t
+
+-- This type describes collections of base types.
+-- Currently collections of collections are not supported.
+data Collection = CNone   { elementType:: Type }   -- not a collection at all
+                | CVector { elementType:: Type }
+                | CSet    { elementType:: Type }
+
+instance Show Collection where
+    show = render . pp
+
+instance PP Collection where
+    pp (CNone t) = pp t
+    pp (CVector t) = "Vector<" <> (pp t) <> ">"
+    pp (CSet t)    = "Set<" <> (pp t) <> ">"
+
+-- resolves a TypeDef to its underlying type (which is a Collection description)
+resolveType :: DatalogProgram -> Type -> Collection
+resolveType d t =
+    -- trace (show t) $
+    case t of
+        -- it is arguably wrong to hardwire these type names in the compiler, but we assume that
+        -- they are more or less standard types.  Maybe the compiler should know about them?
+        TUser{..} -> if typeName == "std.Ref"
+                     then resolveType d $ head typeArgs
+                     else if elem typeName ["std.Set", "tinyset.Set64"] then
+                         let inner = resolveType d $ head typeArgs in
+                         case inner of
+                              CNone elem   -> CSet elem
+                              _            -> error $ "Nested collections not supported: " ++ show t
+                     else if typeName == "std.Vector" then
+                         let inner = resolveType d $ head typeArgs in
+                         case inner of
+                              CNone elem   -> CVector elem
+                              _            -> error $ "Nested collections not supported: " ++ show t
+                     else resolveType d actual
```

Comment:
There's a `typ'` function in `Type.hs` that recursively unwraps typedef's, returning a TOpaque or some other type that is not `TUser`. It might help to simplify this code...